### PR TITLE
Allow defining injection signal

### PR DESCRIPTION
**Pull request by prryplatypus**
Solves #19

### CODE REVIEW

File: sanic_ext/bootstrap.py [hunk 1]
```diff
@@ -128,10 +129,21 @@ def add_dependency(
...
-        self._injection_registry.register(type, constructor)
+
+        if isinstance(signal, str):
+            signal = Event(signal)
+
+        if signal.value not in (
+            "http.routing.after",
+            "http.handler.before",
+        ):
+            raise Exception()  # TODO: Check which exception
```
Comment (ahopkins) [inline sanic_ext/bootstrap.py:144]: `SanicException` works for this

File: sanic_ext/extensions/injection/injector.py [hunk 2]
```diff
@@ -15,6 +15,32 @@
...
     signature_registry = _setup_signature_registry(app, injection_registry)
 
+    @app.before_server_start
+    def register_needed_signals(app: Sanic, _):
+        app.signal_router.reset()
+
+        for signal in injection_registry.signals:
+
+            @app.signal(signal)
+            async def inject_kwargs(request, route, kwargs, **_):
```
Comment (ahopkins) [inline sanic_ext/extensions/injection/injector.py:25]: These args are not going to be the same. I would do:

```python
async def inject_kwargs(request, **_):
```

And then extract everything from the `Request` object.

**Comment by prryplatypus** [comment]
Hey @ahopkins, I'm having an issue with these tests, and I'm guessing that resetting and finalizing the signal router doesn't mix well with the testing client.

The logs can be seen here:
Logs

As you can see from the code here only the first log is output, and the second one isn't while serving a request. Thoughts on this?
Code

**Comment by prryplatypus** [comment]
Managed to work around it in https://github.com/sanic-org/sanic-ext/pull/127/commits/c43dea7af73fb96de9719d3c30c045a0e03c2dab, only we can't register signals dynamically

**Comment by ahopkins** [comment]
@prryplatypus Doesn't look like that will work. Global variables is not the solution.

**Comment by ahopkins** [comment]
What is the problem? Why can we not just register it as a tuple, then have a signal factory that takes the event as an input?

**Comment by prryplatypus** [comment]
I don't think there's any global variables in the workaround!

The issue seems to be that the signals don't get properly registered if it is done from within another signal handler.

As you can see from the logs, the signals technically got registered, but when a request was served, it acted as if no injection signal handlers had been defined.

Therefore what the workaround does is define both signal handlers always and within them only look for injections if there is any defined for their signal they are handling.

**Comment by prryplatypus** [comment]
Per discussion with @ahopkins, we're going to reduce the scope of this PR for now and simply make it a config value that applies to all injections. This can change in the future when we make it easier to dynamically add signal handlers from within other signal handlers.

All changes up to now reverted in https://github.com/sanic-org/sanic-ext/pull/127/commits/1d96f4a3b8deaca85b5318758bb7cca24270f8ee.

**Review by ahopkins — APPROVED** [review]

**Pull request merged**